csrc/optimization/remove_empty.cpp

```diff
@@ -0,0 +1,493 @@
+  void markLive(Statement* stmt) {
+    live_statements_.insert(stmt);
+  }
+
+  //! Ensure that a Statement and its upstream Statements are alive. If it is an
+  //! Expr, ensure all its inputs are alive. If it's a Val with a definition,
+  //! recursive to the definition.
+  void markLiveRecursive(Statement* stmt) {
+    if (isLive(stmt)) {
+      return;
+    }
+    markLive(stmt);
+    if (stmt->isVal() && stmt->asVal()->definition()) {
+      markLiveRecursive(stmt);
+    } else {
+      for (const auto inp : stmt->asExpr()->inputs()) {
+        markLiveRecursive(inp);
+      }
+    }
+  }
+
+  //! Mark a Statement* as dead so that we avoid dereferencing it later
+  void markDead(Statement* stmt) {
+    live_statements_.erase(stmt);
+  }
+
+ private:
+  Fusion* fusion_;
+
+  //! Statements are marked dead by removing them from this set
+  std::unordered_set<Statement*> live_statements_;
+};
+
+//! Get a vector of the integer positions of constant zero extent axes in the
+//! input domain. This will typically be used like
+//! `emptyAxes(TensorDomain::noReductions(tv->getMaybeRFactorDomain()))`
+std::vector<size_t> emptyAxes(std::vector<IterDomain*> domain) {
+  std::vector<size_t> empty_axes;
+  for (auto ax : c10::irange(domain.size())) {
+    auto id = domain.at(ax);
+    if (id->extent()->isConst() && id->extent()->evaluateInt() == 0) {
+      empty_axes.push_back(ax);
+    }
+  }
+  return empty_axes;
+}
+
+//! Check whether a TensorView is empty. During concretization, we traverse to
+//! find a minimal set of TensorViews that have zero extents, and we then set
+//! their extents to a constant 0. Here we check for those constant zero
+//! extents.
+bool isTVEmpty(TensorView* tv) {
+  return !emptyAxes(TensorDomain::noReductions(tv->getMaybeRFactorDomain()))
+              .empty();
+}
+
+//! EmptyTensorRemover performs a backward traversal of the Fusion. When it
+//! detects a TensorView that has at least one extent that is zero, we do the
+//! following:
+//!
+//!   1. If the empty Tensorview is a Fusion output, we replace it with a
+//!   TensorView created by `full` having the same shape. Since the original
+//!   tensor is empty, there is nothing to compute, so this eliminates a branch
+//!   of trivial code.
+//!   2. If the empty TensorView is the input of a `cat` op along the empty
+//!   dimension, we replace the cat op with a new one having the empty input
+//!   removed.
+//!   3. If the empty Tensorview is the input to a ReductionOp or WelfordOp and
+//!   the empty dimensions are reduced, we replace the op with `full` since
+//!   there are no elements being reduced. Note that if any empty axes are not
+//!   reduced, we will not encounter this case since it will have been removed
+//!   earlier in the backward traversal under condition 1.
+//!   4. If the empty Tensorview is the input to a PadOp (which is not input to
+//!   a CatOp) then we replace the pad with `full(pad_value)`.
+//!
```

Comment:
I don't yet handle `MmaOp` here but it is another op that results in a `full` whenever some of the `kAxes` are empty.

Reply:
```python
import torch
A = torch.randn([2, 0])
print(A @ A.T)
# tensor([[0., 0.],
#         [0., 0.]])
```

Reply:
I was thinking I'd just add `MmaOp` support now that I realized it's an empty-removing op.

Reply:
Just pushed a commit that adds `MmaOp` support.